src/util/KinBodyParser.cpp

```diff
@@ -0,0 +1,532 @@
+  }
+  catch(std::exception const& e)
+  {
+    dterr << "[KinBodyParser] LoadFile '" << kinBodyFileUri.toString()
+          << "' Fails: " << e.what() << "\n";
+    return nullptr;
+  }
+
+  return readKinBody(kinBodyDoc, kinBodyFileUri, nullOrRetriever);
+}
+
+namespace {
+
+//==============================================================================
+dart::dynamics::SkeletonPtr readKinBody(
+  tinyxml2::XMLDocument& kinBodyDoc,
+  const dart::common::Uri& baseUri,
+  const dart::common::ResourceRetrieverPtr& nullOrRetriever)
+{
+  const auto retriever = getRetriever(nullOrRetriever);
+
+  // Transform all the element names to lower cases because the KinBody spec is
+  // case-insensitive. We use lower cases to read the XML elements.
+  transformElementNamesToLowerCases(kinBodyDoc);
+
+  auto kinBodyEle = kinBodyDoc.FirstChildElement("kinbody");
+  if (!kinBodyEle)
+  {
+    dterr << "[KinBodyParser] KinBody document does not contain <KinBody> as "
+          << "the root element.\n";
+    return nullptr;
+  }
+
+  return readKinBody(kinBodyEle, baseUri, retriever);
+}
+
+//==============================================================================
+dart::dynamics::SkeletonPtr readKinBody(
+    tinyxml2::XMLElement* kinBodyEle,
+    const dart::common::Uri& baseUri,
+    const dart::common::ResourceRetrieverPtr& retriever)
+{
+  assert(kinBodyEle != nullptr);
+  // kinBodyEle shouldn't be nullptr at all, which always should be
+  // gauranteed because this function is only used internally to be so. If we
+  // find the nullptr case, then we should fix the caller.
+
+  auto skeleton = dart::dynamics::Skeleton::create();
+
+  // Read name attribute.
+  //
+  // If name attribute doesn't exist, getAttributeString() returns empty string.
+  // Empty name for skeleton is allowed.
+  auto name = dart::utils::getAttributeString(kinBodyEle, "name");
+  if (name.empty())
+  {
+    dtwarn << "[KinBodyParser] The <KinBody> in '" << baseUri.toString()
+           << "' doesn't have name attribute or has empty name.\n";
+  }
+  skeleton->setName(name);
+
+  // Get Body node
+  auto bodyEle = kinBodyEle->FirstChildElement("body");
+  if (bodyEle == nullptr)
+  {
+    dterr << "[KinBodyParser] KinBody document '" << baseUri.toString()
+          << "' does not contain <Body> element "
+          << "under <KinBody> element.\n";
+    return nullptr;
+  }
+
+  auto bodyNodeInfo = readBodyNodeInfo(bodyEle, baseUri);
+
+  dart::dynamics::FreeJoint::Properties jointProps;
+  jointProps.mName = DEFAULT_KINBODY_ROOT_JOINTNAME;
+  jointProps.mT_ParentBodyToJoint = bodyNodeInfo.initTransform;
+
+  auto jointAndBodyNode = skeleton->createJointAndBodyNodePair<
+      dart::dynamics::FreeJoint>(nullptr, jointProps, bodyNodeInfo.properties);
+  assert(jointAndBodyNode.first && jointAndBodyNode.second);
+
+  readGeoms(jointAndBodyNode.second, bodyEle, baseUri, retriever);
+  skeleton->resetPositions();
+  skeleton->resetVelocities();
```

Comment:
Nit: Why are these calls to `resetPositions` and `resetVelocities` necessary? We just created `Skeleton` earlier in this function.